ops/testing.py

```diff
@@ -187,6 +187,40 @@ class TestCharm(self._charm_cls):
         TestCharm.__name__ = self._charm_cls.__name__
         self._charm = TestCharm(self._framework)
 
+    def upgrade(self) -> None:
+        """Simulate a charm upgrade.
+
+        This triggers stop, upgrade-charm, config-changed, leader_settings_changed and start hooks.
+        This does NOT trigger a pebble-ready hook (use :meth:`.container_pebble_ready` for that).
+
+        Before calling :meth:`upgrade`, a Charm instance must exist, so you must call
+        :meth:`.begin` or :meth:`.begin_with_initial_hooks` first.
+        """
+        if self._charm is None:
+            raise RuntimeError('cannot call the upgrade method before begin')
+
+        # Reinitialize the charm-under-test
```

Comment:
I think `reinitialize_charm` should be method-worthy on its own (i.e. without emitting `stop` first, and without emitting any events after.

Use cases: https://github.com/PietroPasotti/traefik-route-k8s-operator/blob/main/tests/unit/conftest.py (see last def)